### PR TITLE
Ux 585 listbox improves

### PR DESCRIPTION
**Pull request by nahumzs**
### 🛠 Purpose
Adding two new props to the ListBox *isSelected* and *defaultIsSelected*, given the ability to programmatically control the selected state of the`<ListBox />` in a more friendly way. 

*isSelected*: is a "controlled" prop that sync the internal state of the ListBox with the value passed down. 

*isDefaultSelected* is a uncontrolled prop and describe the initial state of the listbox at the moment of the first render.

---

### ✏️ Notes to Reviewer

The consumer is in charge of the `isSelected` controlled prop which synchronizes the internal state of the `ListBox`. But because a side effect could occur at the moment of selecting one or more options depending on the context, the onClick event still is not controlled, which means that even if the developer declares an onClick event on the Option itself, will not gain control over that event, same goes for up/down key events or the clear button when an option is selected.

Because the previous the developer might need to sync her own controlled state if a changed is fired after one of this event occurred, this can be done via the onChange event.  Ex.

```js
const Options = [options, setOptions] = React.useState([{label: "option 1", isSelected: true}, {label: "option 2", isSelected: false}]
)

function handleClick() {
   // modify the options state to get the desired next state
  setOptions(() => [...yourModifiedArray])
}

function handleChange(selectedOption) {
   console.log("option selected", selectedOption);
   // the developer might want to sync its own state here, but this is an optional step. 
}

<button type="button" onClick={handleClick}>Selected first item</button>
<ListBox onChange={handleChange}>
  {options.map((option, index) => (
       <ListBox.Option key={index} isSelected={option.isSelected}>
          {option.label}
       </ListBox.Option>)
    ) }
</ListBox>
```

---

### 🖥 Screenshots

[Single ListBox programmatically controlling `isSelected` state.](http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves/?path=/story/listbox-single--controlled-isselected-listbox)

[Multi ListBox programmatically controlling `isSelected` state.](http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves/?path=/story/listbox-multi--controlled-listbox)

[Single ListBox programmatically using `defaultIsSelected` prop.](http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves/?path=/story/listbox-single--uncontrolled-defaultisselected-listbox)

[Multi ListBox programmatically using `defaultIsSelected` prop:](http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves/?path=/story/listbox-multi--uncontrolled-defaultisselected-listbox)

---
 
 ### 📙 Storybook 
 <a href='http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves' target="_blank" rel="noopener">http://storybooks.highbond-s3.com/paprika/UX-585-Listbox-improves</a>

### CODE REVIEW

File: packages/DropdownMenu/test/dropdownMenu.spec.js [hunk 1]
```diff
@@ -80,13 +80,13 @@ describe("DropdownMenu", () => {
...
+    it("should hide all dropdown menus when replacement cancel button is clicked", async () => {
       fireEvent.click(getByText(/cancel/i));
-      setTimeout(() => {
+
+      await wait(() => {
         expect(queryByText(/confirm delete/i)).not.toBeInTheDocument();
         expect(queryByText(/edit/i)).not.toBeInTheDocument();
-        done();
-      }, 350);
+      });
```
Comment (nahumzs) [inline packages/DropdownMenu/test/dropdownMenu.spec.js]: boy scouting cleaning setTimeout from our tests.

File: packages/ListBox/src/components/Option/useIsSelectedOption.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        });
+        return;
+      }
+
+      if (props.isSelected === true) {
+        // we want to execute the method exclusively on the prop with selected === true
+        selectSingleOption({ activeOptionIndex: index, state, dispatch, isOpen: false, onChange: noop });
+      }
+    }
+  }, [dispatch, index, props.isSelected]); // eslint-disable-line
```
Comment (nahumzs) [inline packages/ListBox/src/components/Option/useIsSelectedOption.js]: I'm removing state from the deps if not this hook will keep firing all the time.

File: packages/ListBox/src/hooks/hooks.js [hunk 3]
```diff
@@ -93,7 +93,9 @@ export function useOnScrolled() {
   React.useLayoutEffect(() => {
     if (!state.refListBox.current) return;
 
-    if (state.isOpen && state.options[state.activeOption] && state.shouldContentScroll) {
+    const shallScrolled = (state.shouldContentScroll && state.isInline) || (state.isOpen && state.shouldContentScroll);
```
Comment (nahumzs) [inline packages/ListBox/src/hooks/hooks.js]: the scroll was not working correctly if the ListBox was inline.

File: packages/ListBox/src/components/Divider/Divider.js [hunk 4]
```diff
@@ -4,9 +4,9 @@ import { dividerCSS } from "./Divider.styles";
 
 const propTypes = {
   // isDisable is use internally as a default prop
-  // the prop is read by the option/helpers/optionState.js which assigned in the store
-  // later can be ignore by the keyup keydown event seens is consider disable see options/helpers/options.js
-  // function getNextOptionActiveIndex
+  // the prop is read by the option/helpers/optionState.js which is assigned in the store
+  // it helps to ingore the divider while using the keyboard.
```
Comment (alexzherdev) [inline packages/ListBox/src/components/Divider/Divider.js]:
```suggestion
  // it helps to ignore the divider while using the keyboard.
```

File: packages/ListBox/src/hooks/hooks.js (hunk 3 above)
Comment (alexzherdev) [inline packages/ListBox/src/hooks/hooks.js]:
```suggestion
    const shouldScroll = (state.shouldContentScroll && state.isInline) || (state.isOpen && state.shouldContentScroll);
```
?

File: packages/ListBox/src/store/reducer.js [hunk 5]
```diff
@@ -48,6 +48,32 @@ export default function reducer(state, { type, payload }) {
...
 
     case useListBox.types.selectMultipleOption: {
+      const { activeOptionIndex, isSelected } = payload;
+      let selectedOptions = [];
+
+      if (isSelected) {
+        // remove duplicates
+        selectedOptions = [...new Set([...state.selectedOptions, activeOptionIndex])];
+      } else if (state.selectedOptions.includes(activeOptionIndex)) {
+        const cloneSelectedOptions = state.selectedOptions.slice(0);
```
Comment (alexzherdev) [inline packages/ListBox/src/store/reducer.js]: How about
```suggestion
        selectedOptions = state.selectedOptions.filter(option => option !== activeOptionIndex);
```

File: packages/ListBox/src/hooks/hooks.js [hunk 6]
```diff
@@ -103,10 +105,10 @@ export function useOnScrolled() {
           offsetTop = 0;
         }
 
-        state.refListBox.current.scrollTo(0, offsetTop - 10);
+        if (state.refListBox.current) state.refListBox.current.scrollTo(0, offsetTop - 10);
```
Comment (alexzherdev) [inline packages/ListBox/src/hooks/hooks.js]: Btw I thought `scrollTo` is not supported in IE?

**Comment by nahumzs** [comment]
Polyfill scrollTo added to helper package [polyfill](https://github.com/acl-services/paprika/pull/205/files#diff-b8c75b0d5ddfd0e7356306bcd7e25a91)

File: packages/ListBox/src/hooks/hooks.js (hunk 6 above)
Comment (nahumzs) [reply packages/ListBox/src/hooks/hooks.js]: correct [see](https://github.com/acl-services/paprika/pull/205#issuecomment-535198493), I added a polyfill for Edge, should work on IE11 as well but currently storybook is not working anymore on it.

File: packages/ListBox/src/components/Option/useIsSelectedOption.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import React from "react";
+import useListBox from "../../useListBox";
+import { selectSingleOption, selectMultipleOption } from "../Options/helpers/options";
+
+// previnting to report twice the onChange when props changed.
```
Comment (oscarkwan) [inline packages/ListBox/src/components/Option/useIsSelectedOption.js]: preventing*

**Pull request merged**